drivers/counter/counter_ll_stm32_timer.c

```diff
@@ -0,0 +1,623 @@
+			err = -ETIME;
+		}
+
+		/* Interrupt is triggered always for relative alarm and
+		 * for absolute depending on the flag.
+		 */
+		if (irq_on_late) {
+			counter_stm32_counter_stm32_set_cc_int_pending(dev, id);
+		} else {
+			DEV_CFG(dev)->ch_data[id].callback = NULL;
+		}
+	} else {
+		enable_it[id](timer);
+	}
+
+	return err;
+}
+
+static int counter_stm32_set_alarm(const struct device *dev, uint8_t chan,
+				   const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct counter_stm32_config *config = DEV_CFG(dev);
+	struct counter_stm32_ch_data *chdata = &config->ch_data[chan];
+
+	if (alarm_cfg->ticks >  counter_stm32_get_top_value(dev)) {
+		return -EINVAL;
+	}
+
+	if (chdata->callback) {
+		return -EBUSY;
+	}
+
+	chdata->callback = alarm_cfg->callback;
+	chdata->user_data = alarm_cfg->user_data;
+
+	return counter_stm32_set_cc(dev, chan, alarm_cfg);
+}
+
+static int counter_stm32_cancel_alarm(const struct device *dev, uint8_t chan)
+{
+	const struct counter_stm32_config *config = DEV_CFG(dev);
+
+	disable_it[chan](config->timer);
+	config->ch_data[chan].callback = NULL;
+
+	return 0;
+}
+
+static int counter_stm32_set_top_value(const struct device *dev,
+				       const struct counter_top_cfg *cfg)
+{
+	const struct counter_stm32_config *config = DEV_CFG(dev);
+	TIM_TypeDef *timer = DEV_CFG(dev)->timer;
+	struct counter_stm32_data *data = DEV_DATA(dev);
+	int err = 0;
+
+	for (int i = 0; i < counter_get_num_of_channels(dev); i++) {
+		/* Overflow can be changed only when all alarms are
+		 * disabled.
+		 */
+		if (config->ch_data[i].callback) {
+			return -EBUSY;
+		}
+	}
+
+	LL_TIM_DisableIT_UPDATE(timer);
+	LL_TIM_SetAutoReload(timer, cfg->ticks);
+	LL_TIM_ClearFlag_UPDATE(timer);
+
+	data->top_cb = cfg->callback;
+	data->top_user_data = cfg->user_data;
+
+	if (!(cfg->flags & COUNTER_TOP_CFG_DONT_RESET)) {
+		LL_TIM_SetCounter(timer, 0);
+	} else if (read(dev) >= cfg->ticks) {
+		err = -ETIME;
+		if (cfg->flags & COUNTER_TOP_CFG_RESET_WHEN_LATE) {
+			LL_TIM_SetCounter(timer, 0);
+		}
+	}
+
+	if (cfg->callback) {
+		LL_TIM_EnableIT_UPDATE(timer);
+	}
+
+	return err;
+}
+
+static uint32_t counter_stm32_get_pending_int(const struct device *dev)
+{
+	const struct counter_stm32_config *cfg = dev->config;
+	uint32_t pending = 0;
+
+	switch (counter_get_num_of_channels(dev)) {
+	case 4:
+		pending |= LL_TIM_IsActiveFlag_CC4(cfg->timer);
+	case 3:
+		pending |= LL_TIM_IsActiveFlag_CC3(cfg->timer);
+	case 2:
+		pending |= LL_TIM_IsActiveFlag_CC2(cfg->timer);
+	case 1:
+		pending |= LL_TIM_IsActiveFlag_CC1(cfg->timer);
+	}
+
+	return !!pending;
+}
+
+/**
+ * Obtain timer clock speed.
+ *
+ * @param pclken  Timer clock control subsystem.
+ * @param tim_clk Where computed timer clock will be stored.
+ *
+ * @return 0 on success, error code otherwise.
+ *
+ * This function is ripped from the PWM driver; TODO handle code duplication.
+ */
+static int counter_stm32_get_tim_clk(const struct stm32_pclken *pclken, uint32_t *tim_clk)
```

Comment:
this could be generalized, since it's basically copying PWM code (maybe an internal API?)

Reply:
Yep, for sure. But @ABOSTM suggested this was fine for now, and that we can worry about code duplication in a later PR. Do you feel it should be handled here?